### PR TITLE
Implement using Proxy

### LINKED ISSUE #29: Return ES2015 Proxy object
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/pify/issues/29)

**[ECMAScript6 Proxies](http://www.ecma-international.org/ecma-262/6.0/#sec-proxy-object-internal-methods-and-internal-slots) are currently not available on node, therefore this issue is a reminder for future versions.**

`pify(obj)` should return an ECMAScript6 Proxy object, forwarding almost all operations to `obj`, and promisifying only some functions, thus we almost surely get rid of most non-enumerable and inheritance issues.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $70.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($70.00)
### Submitted pull Requests
- [#73 Implement using Proxy](https://issuehunt.io/r/sindresorhus/pify/pull/73)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/pify/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### LINKED ISSUE #57: Promisifying all methods of a ES6 Class instance doesn't seem to work
The readme example shows it should be possible to promisify all methods in a module.

```
// Promisify all methods in a module
pify(fs).readFile('package.json', 'utf8').then(data => {
	console.log(JSON.parse(data).name);
	//=> 'pify'
});
```

This doesn't seem to work for ES6 class instances:

```
// test.js
const pify = require('pify');

class MyClass {
  myMethod(cb) {
    cb(null, 'hello');
  }
}

const myClass = new MyClass();
// works
myClass.myMethod((err, str) => console.log(str));
// works
pify(myClass.myMethod)().then(data => console.log(data));
// does not work
pify(myClass).myMethod().then(data => console.log(data));
```
```
$ node test.js
hello
D:\sandbox\master-project\test.js:13
pify(myClass).myMethod().then(data => console.log(data));
              ^

TypeError: pify(...).myMethod is not a function
    at Object.<anonymous> (D:\sandbox\master-project\test.js:13:15)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
```

I'm not sure if this is by design or not though. The example clearly says "Promisify all methods in a **module**", so maybe it should only be used for CommonJS modules.

### PR DESCRIPTION
**Pull request by frangio**
Fixes #29. Fixes #57.

This implementation changes the behavior when the wrapped object is mutated. It will act as a promisified "view" of the underlying object, rather than a promisified clone. Thus, if properties (including methods) change in the underlying object, the promisified view will also be updated. This change is reflected in a change in one of the existing tests.

I also fixed a bug where symbol properties were incorrectly handled, although that's unrelated to using Proxy.

Regarding this issue in #32:

> [[Get]] for proxy objects enforces the following invariants: The value reported for a property must be the same as the value of the corresponding target object property if the target object property is a non-writable, non-configurable own data property.

As far as I could tell, addressing this is complicated because you need to get the property descriptor which may be inherited, and there is no native API to do that so you have manually walk the prototype chain.

Where should we document the caveat that pify will not work in this case?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#29: Return ES2015 Proxy object](https://issuehunt.io/repos/41585934/issues/29)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by sindresorhus** [comment]
> As far as I could tell, addressing this is complicated because you need to get the property descriptor which may be inherited, and there is no native API to do that so you have manually walk the prototype chain.

Would be nice to just handle this though. Imagine using `pify` on an object from a package you don't control, then that package makes a property non-writable in a patch release, and suddenly `pify` is crashing. While we could document it, it's hard for users to protect themselves against it.

See https://github.com/sindresorhus/on-change/pull/16 for how it was handled on another package of mine.

> target object property is a non-writable, non-configurable own data property.

Doesn't the "own" word imply it doesn't apply when inherited?

File: test.js [hunk 1]
```diff
@@ -278,3 +256,24 @@ test('promisify prototype function', async t => {
...
+
+	t.is(await pified.foo(), 'new');
+});
+
+test('symbol keys', async t => {
+	const sym = Symbol('sym');
+	const obj = {[sym]: cb => setImmediate(cb)};
+	const pified = m(obj);
+	await pified[sym]();
+	t.pass();
```
Comment (sindresorhus) [inline test.js]: Use `t.notThrows` instead.

**Comment by frangio** [comment]
>Doesn't the "own" word imply it doesn't apply when inherited?

Seems so! I totally missed that.

**Comment by frangio** [comment]
I think #63 was fixed in #69. However, it's good that you brough it up because this PR breaks `bind` when `pify` is used on a function, because it's also promisifying the function's methods. So `pify(fn).bind` is actually a promisified version of `bind`, and `pify(fn).bind(obj)` is not a function but a promise, and throws an error if it's called as a function.

To fix this I can add an additional condition that we will not promisify a property if its value is taken from `Function.prototype`. What do you think about this? It feels ad-hoc but since the library is meant to be used with functions in this way, I think it's justified.

***

#48 is fixed-ish. The problem is that a method called through the pify proxy will receive as `this` the proxy and not the underlying object, and this will likely break the object because internally it doesn't expect to have promisified methods. I think this can be fixed by switching the proxy by its target in `apply`, but it has to be done carefully to not break `bind`. I will look into this.

**Comment by sindresorhus** [comment]
> To fix this I can add an additional condition that we will not promisify a property if its value is taken from Function.prototype. What do you think about this? It feels ad-hoc but since the library is meant to be used with functions in this way, I think it's justified.

👍 

> I think this can be fixed by switching the proxy for its target in apply, but it has to be done carefully to not break bind. I will look into this.

👍

**Comment by frangio** [comment]
@sindresorhus Can you please clarify what tests you’d like me to add?

**Comment by frangio** [comment]
I've tackled the remaining issues mentioned, and added tests for those things. I feel that the existing test suite is quite comprehensive. If you have concrete tests you would like me to add let me know though.

**Comment by sindresorhus** [comment]
Looks good now. Great work 👌

**Pull request merged**